docs/MinimallyImpactfulServicing.md

```diff
@@ -0,0 +1,125 @@
+# Minimally Impactful Servicing of eBPF For Windows
+
+## Overview
+As newer versions of eBPF for Windows are released, it is desirable to allow updating
+the code currently executing on systems, without requiring a reboot. Simply stopping and restarting eBPF is
+not a viable solution as applications are dependent on state stored within eBPF,
+in the form of map, program, and link objects. This document outlines a strategy
+for replacing the currently executing version of eBPF for Windows with a newer
+version.
+
+## Architectural changes
+Under the new architecture, the eBPF for Windows execution context (the portion
+of eBPF hosting the virtual machine in which eBPF programs execute) is split
+into two parts.
+
+The first part hosts the code that interacts with the Windows I/O Manager (the
+part of the Windows kernel responsible for exposing handles to user mode). This
+part of the code cannot be serviced without restarting the applications using
+eBPF due to constraints imposed by the Windows kernel. The expectation is that
+this portion of the code will be relatively stable and will not require frequent
+updates. This part is referred to as eBPF Proxy or the proxy driver.
+
+The second part hosts the code that implements the eBPF for Windows virtual
+machine. This part of the code supports serializing its state in a format that is
+both forward and backward compatible. During termination of this code it
+can optionally serialize its state and during initiation it can optionally
+deserialize its state. The net effect is that this part of the code can be
+reloaded without any user visible change. This part is referred to as eBPF Core
+or the core driver.
+
+To ensure that the state stored in eBPF Core remains consistent across updates,
+the following actions are taken:
+1. Calls from user mode are blocked in eBPF Proxy.
+2. eBPF hook providers are notified of servicing start.
+3. eBPF programs are detached from attach points prior unloading the old code
+and reattached after loading the new code.
+4. eBPF hook providers are notified of servicing end.
+5. Calls from user mode are unblocked in eBPF Proxy.
+
+Each eBPF hook provider is responsible for picking an appropriate strategy for
+handling events during servicing. Possible behaviors include queueing the events
+or dropping them. As an example, the XDP hook might choose to return XDP_DROP
```

Comment:
Does this not count as downtime or an impact to data path?

Reply:
That depends on the extension handles it. The extension could buffer the packets during servicing and complete them when the servicing ends.